rfcs/20210119-determinism.md

```diff
@@ -0,0 +1,82 @@
+# RFC: Enabling Determinism in TensorFlow
+  
+| Status        | Proposed                                             |
+:---------------|:-----------------------------------------------------|
+| **Author(s)** | Pankaj Kanwar (Google), Reed Wanderman-Milne (Google)|
+| **Sponsor**   | Sanjoy Das (Google)                                  |
+| **Updated**   | 2021-01-31                                           |
+
+
+## Objective
+Allow users to enable determinism behavior in TensorFlow. This means if the user runs a TensorFlow program multiple times, the model outputs and weights will be the same each time. Determinism will be supported on CPUs and GPUs.
+ 
+To get deterministic behavior, users must do the following:
+
+* Enable determinism using the API proposed in this doc.
+* Use same hardware in every run.
+* Use the same software environment every run (OS, checkpoints, version of TF, environmental variables, etc).
+* Not use constructs outside TensorFlow that are nondeterministic, such as Python’s `random` module or using multiple threads/processes in ways that influence TensorFlow’s behavior.
```

Comment:
Python's `random` module is not nondeterministic. As far as I know, it supports full deterministic functionality, but to enable determinism the user just has to initialize it (as she would for any other compute process from which she desired deterministic functionality).

Suggested change: Not use constructs outside TensorFlow that are nondeterministic, such as Python’s `random` module (without appropriate use of its ability to be initialized/seeded) or using multiple threads/processes in ways that influence TensorFlow’s behavior.